pkg/slices/slices_test.go

```diff
@@ -0,0 +1,163 @@
+	},
+	{
+		name:     "all uniques",
+		input:    []int{1, 3, 4, 2, 9, 7, 6, 10, 5, 8},
+		expected: []int{1, 3, 4, 2, 9, 7, 6, 10, 5, 8},
+	},
+	{
+		name:     "all duplicates",
+		input:    []int{1, 1, 1, 1, 1, 1, 1, 1, 1, 1},
+		expected: []int{1},
+	},
+	{
+		name:     "uniques and duplicates",
+		input:    []int{1, 2, 2, 1, 1, 3, 1, 3, 1, 4},
+		expected: []int{1, 2, 3, 4},
+	},
+}
+
+func TestUnique(t *testing.T) {
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			got := Unique(tc.input)
+			assert.ElementsMatch(t, tc.expected, got)
+		})
+	}
+}
+
+func TestSortedUnique(t *testing.T) {
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			got := SortedUnique(tc.input)
+			assert.ElementsMatch(t, tc.expected, got)
+		})
+	}
+}
+
+func TestSortedUniqueFunc(t *testing.T) {
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			got := SortedUniqueFunc(
+				tc.input,
+				func(i, j int) bool {
+					return tc.input[i] < tc.input[j]
+				},
+				func(a, b int) bool {
+					return a == b
+				},
+			)
+			assert.ElementsMatch(t, tc.expected, got)
+		})
+	}
+}
+
+func TestUniqueKeepOrdering(t *testing.T) {
+	input := []string{"test-4", "test-1", "test-3", "test-4", "test-4", "test-3", "test-5"}
+	expected := []*string{&input[0], &input[1], &input[2], &input[3]}
+
+	got := Unique(input)
+
+	if len(got) != len(expected) {
+		t.Fatalf("expected slice of %d elements, got %d", len(expected), len(got))
+	}
+
+	for i := 0; i < len(expected); i++ {
+		if got[i] != *expected[i] {
+			t.Fatalf("expected value %q at index %d, got %q", *expected[i], i, got[i])
+		}
+
+		if &got[i] != expected[i] {
+			t.Fatalf("expected address of value at index %d to be %x, got %x", i, expected[i], &got[i])
+		}
+	}
+}
+
+// BenchmarkUnique runs the Unique function on a slice of size elements, where each element
+// has a probability of 20% of being a duplicate.
+// At each iteration the slice is restored to its original status and reshuffled, in order
+// to benchmark the average time needed to deduplicate the elements whatever their specific order.
+//
+// This benchmark has been used to experimentally derive the size limit for Unique beyond which
+// the algorithm changes from a O(N^2) search to a map based approach.
+// Forcing Unique to rely only on a single algorithm at a time and running the benchmark with count=5,
+// the compared results extracted with benchstat are the following:
+//
+// name          old time/op    new time/op    delta
+// Unique/96-8     3.17µs ± 9%    4.83µs ±15%  +52.50%  (p=0.008 n=5+5)
+// Unique/128-8    4.97µs ± 5%    5.95µs ± 2%  +19.83%  (p=0.008 n=5+5)
+// Unique/160-8    7.20µs ±12%    7.33µs ± 1%     ~     (p=0.690 n=5+5)
+// Unique/192-8    9.29µs ± 3%    9.07µs ± 2%     ~     (p=0.151 n=5+5)
+// Unique/256-8    15.4µs ± 4%    11.2µs ± 2%  -27.56%  (p=0.008 n=5+5)
+// ...
+//
+// After 192 elements, the map based approach becomes more efficient.
```

Comment:
Can you include results for memory just so we can get an overall sense of the performance?

Reply:
Added memory results, too.
I pointed out in the comment that, since the number of allocations from the loop algorithm is always 0, benchstat reports `+Inf` when comparing the two. But the info in the previous two columns (old alloc/op and new alloc/op) should be enough to highlight the allocations performed by the map approach:

```
...

name          old alloc/op   new alloc/op   delta
Unique/96-8      0.00B       1474.00B ± 2%    +Inf%  (p=0.008 n=5+5)
Unique/128-8     0.00B       3100.00B ± 0%    +Inf%  (p=0.008 n=5+5)
Unique/160-8     0.00B       3113.20B ± 0%    +Inf%  (p=0.008 n=5+5)
Unique/192-8     0.00B       3143.20B ± 0%    +Inf%  (p=0.008 n=5+5)
Unique/256-8     0.00B       6178.00B ± 0%    +Inf%  (p=0.008 n=5+5)

name          old allocs/op  new allocs/op  delta
Unique/96-8       0.00           3.20 ±38%    +Inf%  (p=0.008 n=5+5)
Unique/128-8      0.00           2.00 ± 0%    +Inf%  (p=0.008 n=5+5)
Unique/160-8      0.00           3.00 ± 0%    +Inf%  (p=0.016 n=5+4)
Unique/192-8      0.00           4.00 ± 0%    +Inf%  (p=0.008 n=5+5)
Unique/256-8      0.00           2.00 ± 0%    +Inf%  (p=0.008 n=5+5)
```